[Navigation API] Add test verifying main frame navigation.entries() excludes iframe history.

basuke opened the pull request:
#### ce6ba9ee81f8e3f9117dfeca0d166b0956f2c61b
<pre>
[Navigation API] Add test verifying main frame navigation.entries() excludes iframe history.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302354">https://bugs.webkit.org/show_bug.cgi?id=302354</a>
<a href="https://rdar.apple.com/156914269">rdar://156914269</a>

Reviewed by Rupin Mittal.

Add a test to verify that the main frame&apos;s navigation.entries() correctly excludes history entries
from iframes, as requested in review feedback for 297926@main:
- <a href="https://github.com/WebKit/WebKit/pull/48449#discussion_r2227034292">https://github.com/WebKit/WebKit/pull/48449#discussion_r2227034292</a>

The test performs independent navigations in both the main frame and an iframe, then explicitly
verifies that:
- Main frame entries contain only main frame URLs
- Iframe entries contain only iframe URLs
- Entry counts are independent (not combined)
- No cross-contamination occurs between frame histories

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-mainframe-with-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-mainframe-with-iframe.html: Added.

Canonical link: <a href="https://commits.webkit.org/303021@main">https://commits.webkit.org/303021@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f023bee2737d81a3471dc4658bb808bf5e9b4aa9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3218 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41853 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138322 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/44f06f76-d86a-40d1-920f-52cd0f3444bb) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3200 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3060 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/138322 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6e3b70ef-050e-473f-8996-7ec05bca1ee1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133840 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2310 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/41853 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/138322 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a96914d4-bc5b-4886-9146-b08ddb580343) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2227 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/41853 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81581 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110817 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/41853 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140802 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2961 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/3060 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/140802 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3007 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/41853 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/140802 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27533 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2264 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/41853 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55998 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3029 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2850 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3050 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2959 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-mainframe-with-iframe.html:
Can this be a WPT that we export rather than a webkit only one?

basuke replied:
Sure. That's my plan. Before that, I'm testing with webkit. Should I try landing this from the beginning?

basuke replied:
Created. https://github.com/web-platform-tests/wpt/pull/55988

nt1m replied:
Could we move it to imported/w3c/ in the webkit repo?

basuke replied:
Ah, okay. Is that what you mean? Should I close the upstream PR then?

nt1m replied:
I meant both :)

nt1m reviewed (commented):
I'll let Rupin review the core of the test

RupinMittal commented on LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-mainframe-with-iframe.html:
```diff
@@ -0,0 +1,75 @@
+  assert_equals(navigation.entries().length, start_length + 2, "after #iframe2 outer entries() length");
+  assert_equals(i.contentWindow.navigation.entries().length, 3, "after #iframe2 iframe entries() length");
+
+  // Verify main frame entries contain only main frame URLs
```
We can condense all the checks below this line into just:

```
  // Verify main frame entries contain only main frame URLs
  let mainEntries = navigation.entries().slice(start_index);
  assert_equals(mainEntries.length, 3, "main frame should have 3 entries");

  assert_true(mainEntries[0].url.includes(initialMainURL.split('#')[0]), "main entry 0 should be main frame URL");
  assert_true(mainEntries[1].url.endsWith("#main1"), "main entry 1 should end with #main1");
  assert_true(mainEntries[2].url.endsWith("#main2"), "main entry 2 should end with #main2");

  for (let entry of mainEntries)
    assert_false(entry.url.includes("#iframe"), "main frame entry should not contain iframe hash: " + entry.url);

  // Verify iframe entries contain only iframe URLs
  let iframeEntries = i.contentWindow.navigation.entries();
  assert_equals(iframeEntries.length, 3, "iframe should have 3 entries");

  assert_true(iframeEntries[0].url.includes(initialIframeURL.split('#')[0]), "iframe entry 0 should be iframe URL");
  assert_true(iframeEntries[1].url.endsWith("#iframe1"), "iframe entry 1 should end with #iframe1");
  assert_true(iframeEntries[2].url.endsWith("#iframe2"), "iframe entry 1 should end with #iframe2");

  for (let entry of iframeEntries)
    assert_false(entry.url.includes("#main"), "iframe entry should not contain main frame hash: " + entry.url);
 ```

RupinMittal reviewed (approved):
The logic looks good, but there is some duplication we should condense before landing.

basuke committed:
[Navigation API] Add test verifying main frame navigation.entries() excludes iframe history.
https://bugs.webkit.org/show_bug.cgi?id=302354
rdar://156914269

Reviewed by Rupin Mittal.

Add a test to verify that the main frame's navigation.entries() correctly excludes history entries
from iframes, as requested in review feedback for 297926@main:
- https://github.com/WebKit/WebKit/pull/48449#discussion_r2227034292

The test performs independent navigations in both the main frame and an iframe, then explicitly
verifies that:
- Main frame entries contain only main frame URLs
- Iframe entries contain only iframe URLs
- Entry counts are independent (not combined)
- No cross-contamination occurs between frame histories

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-mainframe-with-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-mainframe-with-iframe.html: Added.

Canonical link: https://commits.webkit.org/303021@main
